pkg/clustermesh/clustermesh.go

```diff
@@ -179,21 +179,31 @@ func (cm *ClusterMesh) NumReadyClusters() int {
 	return cm.common.NumReadyClusters()
 }
 
-// ClustersSynced returns after all clusters were synchronized with the bpf
-// datapath.
-func (cm *ClusterMesh) ClustersSynced(ctx context.Context) error {
-	swgs := make([]*lock.StoppableWaitGroup, 0)
+// SyncedWaitFn is the type of a function to wait for the initial synchronization
+// of a given resource type from all remote clusters.
+type SyncedWaitFn func(ctx context.Context) error
+
+// ServicesSynced returns after that the initial list of shared services has been
+// received from all remote clusters, and synchronized with the BPF datapath.
+func (cm *ClusterMesh) ServicesSynced(ctx context.Context) error {
+	return cm.synced(ctx, func(rc *remoteCluster) SyncedWaitFn { return rc.synced.Services })
+}
+
+func (cm *ClusterMesh) synced(ctx context.Context, toWaitFn func(*remoteCluster) SyncedWaitFn) error {
+	waiters := make([]SyncedWaitFn, 0)
 	cm.common.ForEachRemoteCluster(func(rci common.RemoteCluster) error {
 		rc := rci.(*remoteCluster)
-		swgs = append(swgs, rc.swg)
+		waiters = append(waiters, toWaitFn(rc))
 		return nil
 	})
 
-	for _, swg := range swgs {
-		select {
-		case <-swg.WaitChannel():
-		case <-ctx.Done():
-			return ctx.Err()
+	for _, wait := range waiters {
+		err := wait(ctx)
+
+		// Ignore the error in case the given cluster was disconnected in
+		// the meanwhile, as we do not longer care about it.
```

Comment:
I didn't add it because we already have a similar log output on cluster disconnection (which in this context refers to a user manually removing the cluster configuration, not the disconnection from etcd): https://github.com/cilium/cilium/blob/8e55b7ec3d6e9d1eb692c619c71e42ec5412d33a/pkg/clustermesh/common/remote_cluster.go#L392-L397